raft/heartbeat_manager: timeout memory leak fix

joe-redpanda opened the pull request:
fixes https://redpandadata.atlassian.net/browse/CORE-10326

Changes the logic of lightweight heartbeats to consider the number of failed heartbeats. On enough failed heartbeats hb_m will now follow the full heartbeat pathway which resets the connection.

This fixes a memory leak where if a node was suspended, rpcs would accumulate in the transport buffer to suspended node. Even in an idle cluster, heartbeat rpcs will continuously accumulate in the message queue. Resetting the connection dumps the buffers, freeing all timed out rpcs.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [x] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=26737) <br/>All committers have signed the CLA.

Commit:
raft/heartbeat_manager: timeout memory leak fix

Changes the logic of lightweight heartbeats to consider the number of
failed heartbeats. On enough failed heartbeats hb_m will now follow the
full heartbeat pathway which resets the connection.

This fixes a memory leak where if a node was suspended, rpcs would
accumulate in the transport buffer to suspended node. Even in an idle
cluster, heartbeat rpcs will continuously accumulate in the message
queue. Resetting the connection dumps the buffers, freeing all timed out
rpcs.

bharathv commented on src/v/raft/heartbeat_manager.cc:
```diff
@@ -212,6 +210,17 @@ bool heartbeat_manager::needs_full_heartbeat(
            || leader_flushed_offset != f_meta.last_flushed_log_index;
 }
 
+bool heartbeat_manager::should_lw_heartbeat(
```
nit: wondering if we should just populate reconnect nodes in the lw heartbeats path as well? Its a bit odd that we switch to full heartbeat if nodes needs a reconnect

joe-redpanda replied:
I viewed this as 'something went wrong, lets do a more rigorous check'

joe-redpanda replied:
I have no strong opinions, though

bharathv replied:
ya that seems fair too.

bharathv commented:
can you please add a link to the jira in the description, thanks.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--8976b205-a88c-4907-aae7-7fac8fa38e05-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68607">build#68607</a></summary>


| test_class        | test_method         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------|:--------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterConfigTest | test_valid_settings |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68607#0197ebc9-7d6c-4b9f-8b91-8a6f5e4223b8 | FLAKY         | 17/21    | upstream reliability is '83.77192982456141'. current run reliability is '80.95238095238095'. drift is 2.81955 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

dotnwat commented on src/v/raft/heartbeat_manager.cc:
```diff
@@ -115,10 +115,8 @@ heartbeat_manager::requests_for_range() {
-                _enable_lw_heartbeat()
-                && !needs_full_heartbeat(
-                  follower_metadata, raft_metadata, r->flushed_offset())) {
+              if (should_lw_heartbeat(
+                    r, follower_metadata, raft_metadata, r->flushed_offset())) {
```
is it fair to say that this isn't a real memory leak, but rather a buffer with uncontrolled growth?
